File: learnbot-dsl/LearnBotClient.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys, traceback, Ice, os, math, time, json, ast, copy, threading
import json
import cv2
import urllib
from collections import namedtuple
import numpy as np

ROBOCOMP = ''
try:
	ROBOCOMP = os.environ['ROBOCOMP']
except KeyError:
	print '$ROBOCOMP environment variable not set, using the default value /opt/robocomp'
	ROBOCOMP = '/opt/robocomp'

preStr = "-I/opt/robocomp/interfaces/ -I"+ROBOCOMP+"/interfaces/ --all /opt/robocomp/interfaces/"

Ice.loadSlice(preStr+"RGBD.ice")
Ice.loadSlice(preStr+"Laser.ice")
Ice.loadSlice(preStr+"DifferentialRobot.ice")
Ice.loadSlice(preStr+"JointMotor.ice")
Ice.loadSlice(preStr+"GenericBase.ice")

import RoboCompRGBD
import RoboCompLaser
import RoboCompDifferentialRobot
import RoboCompJointMotor
import RoboCompGenericBase


import signal
signal.signal(signal.SIGINT, signal.SIG_DFL)

ic = None


		
class Client(Ice.Application, threading.Thread):	
	def __init__(self, argv):
		threading.Thread.__init__(self)

		self.adv = 0
		self.rot = 0
		self.max_rot= 0.4
		self.image = np.zeros((240,320,3), np.uint8)
		self.usList = {'front':1000, 'right':1000, 'left':1000, 'back':1000}

	  	global ic

		params = copy.deepcopy(sys.argv)
		if len(params) > 1:
			if not params[1].startswith('--Ice.Config='):
				params[1] = '--Ice.Config=' + params[1]
		elif len(params) == 1:
			params.append('--Ice.Config=config')
		ic = Ice.initialize(params)


		status = 0
		try:
			# Remote object connection for DifferentialRobot
			try:
				proxyString = ic.getProperties().getProperty('DifferentialRobotProxy')
				try:
					basePrx = ic.stringToProxy(proxyString)
					self.differentialrobot_proxy = RoboCompDifferentialRobot.DifferentialRobotPrx.checkedCast(basePrx)
				except Ice.Exception:
					print 'Cannot connect to the remote object (DifferentialRobot)', proxyString
					sys.exit(1)
			except Ice.Exception, e:
				print e
				print 'Cannot get DifferentialRobotProxy property.'
				sys.exit(1)
			# Remote object connection for Laser1
			try:
				proxyString = ic.getProperties().getProperty('Laser1Proxy')
				try:
					basePrx = ic.stringToProxy(proxyString)
					self.laser1_proxy = RoboCompLaser.LaserPrx.checkedCast(basePrx)
				except Ice.Exception:
					print 'Cannot connect to the remote object (Laser)', proxyString
					sys.exit(1)
			except Ice.Exception, e:
				print e
				print 'Cannot get Laser1Proxy property.'
				sys.exit(1)
			# Remote object connection for Laser2
			try:
				proxyString = ic.getProperties().getProperty('Laser2Proxy')
				try:
					basePrx = ic.stringToProxy(proxyString)
					self.laser2_proxy = RoboCompLaser.LaserPrx.checkedCast(basePrx)
				except Ice.Exception:
					print 'Cannot connect to the remote object (Laser)', proxyString
					sys.exit(1)
			except Ice.Exception, e:
				print e
				print 'Cannot get Laser2Proxy property.'
				sys.exit(1)
			# Remote object connection for Laser3
			try:
				proxyString = ic.getProperties().getProperty('Laser3Proxy')
				try:
					basePrx = ic.stringToProxy(proxyString)
					self.laser3_proxy = RoboCompLaser.LaserPrx.checkedCast(basePrx)
				except Ice.Exception:
					print 'Cannot connect to the remote object (Laser)', proxyString
					sys.exit(1)
			except Ice.Exception, e:
				print e
				print 'Cannot get Laser3Proxy property.'
				sys.exit(1)
			# Remote object connection for Laser4
			try:
				proxyString = ic.getProperties().getProperty('Laser4Proxy')
				try:
					basePrx = ic.stringToProxy(proxyString)
					self.laser4_proxy = RoboCompLaser.LaserPrx.checkedCast(basePrx)
				except Ice.Exception:
					print 'Cannot connect to the remote object (Laser)', proxyString
					sys.exit(1)
			except Ice.Exception, e:
				print e
				print 'Cannot get Laser4Proxy property.'
				sys.exit(1)
			# Remote object connection for RGBD
			try:
				proxyString = ic.getProperties().getProperty('RGBDProxy')
				try:
					basePrx = ic.stringToProxy(proxyString)
					self.rgbd_proxy = RoboCompRGBD.RGBDPrx.checkedCast(basePrx)
				except Ice.Exception:
					print 'Cannot connect to the remote object (RGBD)', proxyString
					sys.exit(1)
			except Ice.Exception, e:
				print e
				print 'Cannot get RGBDProxy property.'
				sys.exit(1)




		except:
				traceback.print_exc()
				sys.exit(1)	 

		self.active = True
		self.start()

		
	def run(self):
		while self.active:
			try:
				self.color, self.depth, self.headState, self.baseState = self.rgbd_proxy.getData()
				if (len(self.color) == 0) or (len(self.depth) == 0):
	                                print 'Error retrieving images!'
	                except Ice.Exception:
	                        traceback.print_exc()		
	
			self.image = np.fromstring(self.color, dtype=np.uint8).reshape((240, 320, 3))

			self.readSonars()

			time.sleep(0.01)

	def readSonars(self):

		l1data = self.laser1_proxy.getLaserData()
		minD1 = l1data[0].dist
		for data in l1data:
			if minD1 > data.dist:
				minD1 = data.dist
	
		self.usList["front"] = minD1

		l2data = self.laser2_proxy.getLaserData()
		minD2 = l2data[0].dist
		for data in l2data:
			if minD2 > data.dist:
				minD2 = data.dist

		self.usList["right"] = minD2


		l3data = self.laser3_proxy.getLaserData()
		minD3 = l3data[0].dist
		for data in l3data:
			if minD3 > data.dist:
				minD3 = data.dist

		self.usList["left"] = minD2


		l4data = self.laser4_proxy.getLaserData()
		minD4 = l4data[0].dist
		for data in l4data:
			if minD4 > data.dist:
				minD4 = data.dist

		self.usList["back"] = minD2



	def getSonars(self):
		return self.usList

	def getImage(self):
		return self.image

	def getPose(self):
		x, y, alpha = self.differentialrobot_proxy.getBasePose()	 
		return x, y, alpha
	 	     
	def setRobotSpeed(self, vAdvance=0, vRotation=0):
		if vAdvance!=0 or vRotation!=0:
			self.adv = vAdvance
			self.rot = vRotation
		self.differentialrobot_proxy.setSpeedBase(self.adv,self.rot)	 
				

	def __del__(self):
        	self.active = False


	      
	

File: learnbot_dsl/LearnBotClient.cpp
#include "LearnBotClient.h"


void readImage(params_image * params)
{
  while (true)
  {
    try
    {
      RoboCompRGBD::ColorSeq colorseq;
      RoboCompRGBD::DepthSeq depthseq;
      RoboCompGenericBase::TBaseState bState;
      RoboCompJointMotor::MotorStateMap hState;
      params->rgbd_proxy->getRGB(colorseq, hState, bState);
      params->image_mutex->lock();
      memcpy(params->image->data , &colorseq[0], 320*240*3);
      params->image_mutex->unlock();
    }
    catch(const Ice::Exception &e)
    {
      std::cout << "Error reading form RGBD " << e << std::endl;
    }
  }
};

void readSonars(params_sonars *params)
{
  while (true) {
    int i=0;
    std::vector<LaserPrx> l_prx_vec = params->laser_proxy_vector;
    for (auto lprx:l_prx_vec)
    {
      auto ldata = lprx->getLaserData();
      params->laser_mutex->lock();
      float min=ldata[0].dist;
      for (auto value:ldata)
        if (min<value.dist)
          min = value.dist;
      params->Ldata[i]= min;
      params->laser_mutex->unlock();
      i++;
    }
  }
}
//
// ClientCpp::ClientCpp(const ClientCpp &client)
// {
//   adv = client.adv;
//   rot = client.rot;
//   max_rot = client.max_rot;
//   image = client.image;
//   image_mutex = client.image_mutex;
//   ic = client.ic;
//   differentialrobot_proxy = client.differentialrobot_proxy;
//   rgbd_proxy = client.rgbd_proxy;
//   args_sonars = client.args_sonars;
// }

ClientCpp::ClientCpp(std::vector<std::string> argv_)
{
  char* argv[argv_.size()];
  for (int i=0;i< argv_.size(); i++)
  {
    argv[i] = new char[argv_[i].size()];
    strncpy(argv[i], argv_[i].c_str(), argv_[i].size());
  }
  initialize(argv_.size(), argv);
};

ClientCpp::ClientCpp(int argc, char* argv[])
{
  initialize(argc, argv);
};

void ClientCpp::initialize(int argc, char* argv[])
{
  image_mutex = new std::mutex();
  args_sonars.laser_mutex = new std::mutex();
  if (argc > 1)
  {
    if (!boost::starts_with(argv[1], "--Ice.Config="))
    {
      std::string ll = std::string("--Ice.Config=") + argv[1];
      argv[1] = new char[ll.size()];
      strncpy(argv[1], ll.c_str(), ll.size());
      ic = Ice::initialize(argc, argv);
    }
  }
  else if(argc == 1)
  {
    char* params1[2];
    std::string a1 = argv[0];
    params1[0] = new char[a1.size()];
    strncpy(params1[1], a1.c_str(), a1.size());
    std::string ll = std::string("--Ice.Config=config");
    params1[1] = new char[ll.size()];
    strncpy(params1[1], ll.c_str(), ll.size());
    ic = Ice::initialize(argc, params1);
  }

  std::string proxyString =ic->getProperties()->getProperty("DifferentialRobotProxy");
  differentialrobot_proxy = DifferentialRobotPrx::uncheckedCast( ic->stringToProxy( proxyString ) );
  if(!differentialrobot_proxy) throw std::runtime_error("Invalid proxy DifferentialRobotProxy");


  proxyString = ic->getProperties()->getProperty("RGBDProxy");
  rgbd_proxy = RGBDPrx::uncheckedCast( ic->stringToProxy( proxyString ) );
  if(!rgbd_proxy) throw std::runtime_error("Invalid proxy RGBDProxy");


  image.create(240,320, CV_8UC3);
  params_image args_image;
  args_image.image = &image;
  args_image.image_mutex = image_mutex;
  args_image.rgbd_proxy = rgbd_proxy;
  std::thread t1(readImage, &args_image);
  std::thread t2(readSonars, &args_sonars);
};


std::vector<float> ClientCpp::getSonars()
{
  args_sonars.laser_mutex->lock();
  std::vector<float> ldata;
  for (auto value:args_sonars.Ldata)
    ldata.push_back(value);
  args_sonars.laser_mutex->unlock();
  return ldata;
};


cv::Mat ClientCpp::getImage()
{
  image_mutex->lock();
  return image;
  image_mutex->unlock();
};

std::vector<float> ClientCpp::getPose()
{
  int x, y;
  float alpha;
  differentialrobot_proxy->getBasePose(x,y, alpha);
  std::vector<float> data = {(float)x, (float)y, alpha};
  return data;
};

void ClientCpp::setRobotSpeed(int vAdvance, int vRotation)
{
  if (vAdvance!=0 || vRotation!=0)
  {
      adv = vAdvance;
      rot = vRotation;
  }
  differentialrobot_proxy->setSpeedBase(adv,rot);
};
